### PR TITLE
added radius_settings to aws_directory_service_directory

### LINKED ISSUE #12639: Add Radius support for aws_directory_service_directory
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In our modern security environment, multi-factor authentication on enterprise applications is rapidly becoming a must-have. In order to implement MFA for Amazon's enterprise applications (WorkSpaces, WorkDocs, WorkMail, QuickSight, SSO, and console), it is necessary to configure Radius authentication on `aws_directory_service_directory` resources of type `ADConnector` or `MicrosoftAD`.

### New or Affected Resource(s)

* aws_directory_service_directory

### Potential Terraform Configuration

```hcl
resource "aws_directory_service_directory" "example" {
  name = var.adc_domain
  password = var.adc_pass
  size = "Small"
  type = "ADConnector"
  radius {
    protocol = "PAP"
    label = "MyRadius"
    port = 1200
    retries = 2
    servers = ["172.168.111.12"]
    timeout = 1
    secret = "12345678"
    same_username = true
  }
}
```

### References

* https://docs.aws.amazon.com/directoryservice/latest/admin-guide/ms_ad_mfa.html
* https://docs.aws.amazon.com/directoryservice/latest/admin-guide/ad_connector_mfa.html

### PR DESCRIPTION
**Pull request by apagliara**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12639

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Added Radius MFA support for aws_directory_service_directory
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSDirectoryServiceDirectory'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSDirectoryServiceDirectory -timeout 120m
=== RUN   TestAccAWSDirectoryServiceDirectory_basic
=== PAUSE TestAccAWSDirectoryServiceDirectory_basic
=== RUN   TestAccAWSDirectoryServiceDirectory_tags
=== PAUSE TestAccAWSDirectoryServiceDirectory_tags
=== RUN   TestAccAWSDirectoryServiceDirectory_microsoft
=== PAUSE TestAccAWSDirectoryServiceDirectory_microsoft
=== RUN   TestAccAWSDirectoryServiceDirectory_microsoftStandard
=== PAUSE TestAccAWSDirectoryServiceDirectory_microsoftStandard
=== RUN   TestAccAWSDirectoryServiceDirectory_connector
=== PAUSE TestAccAWSDirectoryServiceDirectory_connector
=== RUN   TestAccAWSDirectoryServiceDirectory_withAliasAndSso
=== PAUSE TestAccAWSDirectoryServiceDirectory_withAliasAndSso
=== RUN   TestAccAWSDirectoryServiceDirectory_microsoftStandard_withRadiusMfa
=== PAUSE TestAccAWSDirectoryServiceDirectory_microsoftStandard_withRadiusMfa
=== RUN   TestAccAWSDirectoryServiceDirectory_disappears
=== PAUSE TestAccAWSDirectoryServiceDirectory_disappears
=== CONT  TestAccAWSDirectoryServiceDirectory_basic
=== CONT  TestAccAWSDirectoryServiceDirectory_microsoft
=== CONT  TestAccAWSDirectoryServiceDirectory_disappears
=== CONT  TestAccAWSDirectoryServiceDirectory_connector
=== CONT  TestAccAWSDirectoryServiceDirectory_microsoftStandard_withRadiusMfa
=== CONT  TestAccAWSDirectoryServiceDirectory_withAliasAndSso
=== CONT  TestAccAWSDirectoryServiceDirectory_tags
=== CONT  TestAccAWSDirectoryServiceDirectory_microsoftStandard
--- PASS: TestAccAWSDirectoryServiceDirectory_basic (719.78s)
--- PASS: TestAccAWSDirectoryServiceDirectory_tags (734.49s)
--- PASS: TestAccAWSDirectoryServiceDirectory_withAliasAndSso (749.65s)
--- PASS: TestAccAWSDirectoryServiceDirectory_disappears (976.49s)
--- PASS: TestAccAWSDirectoryServiceDirectory_connector (1225.98s)
--- PASS: TestAccAWSDirectoryServiceDirectory_microsoftStandard (1816.99s)
--- PASS: TestAccAWSDirectoryServiceDirectory_microsoft (1836.59s)
--- PASS: TestAccAWSDirectoryServiceDirectory_microsoftStandard_withRadiusMfa (1906.25s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1908.528s
```

### CODE REVIEW

**Comment by jgiannuzzi** [comment]
Pinging @breathingdust @anGie44 @bflad — could you please review this PR, or assign someone to review it?

Thanks!

**Comment by breathingdust** [comment]
Hi! 👋 

Thanks for your contribution and for getting in touch @jgiannuzzi . At this time this feature doesn’t have enough community support for the internal team to prioritize. We will continue to monitor the community interest in the feature, and once enough support is demonstrated, the team will queue this PR up for review.

**Comment by tom-henderson** [comment]
Hi @breathingdust, we're really interested in seeing this one merged. Could you please consider queueing it up for review?

**Comment by stackedsax** [comment]
@breathingdust just curious how you measure community support.  @tom-henderson and I would both like to see this prioritized higher -- what else should we do to engage the community to judge whether it's something that folks actually need/want?

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:191]: ValidateFunc: validation.StringInSlice(directoryservice.RadiusAuthenticationProtocol_Values(), false),

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:199]: add length validation https://docs.aws.amazon.com/directoryservice/latest/devguide/API_RadiusSettings.html

`ValidateFunc: validation.StringLenBetween(1, 64),`

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:206]: add validation https://docs.aws.amazon.com/directoryservice/latest/devguide/API_RadiusSettings.html

ValidateFunc: validation.IntBetween(0, 10),

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:216]: add validation https://docs.aws.amazon.com/directoryservice/latest/devguide/API_RadiusSettings.html

ValidateFunc: validation.IntBetween(1, 20),

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:220]: add length validation https://docs.aws.amazon.com/directoryservice/latest/devguide/API_RadiusSettings.html

ValidateFunc: validation.StringLenBetween(8, 512),

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:213]: add validation https://docs.aws.amazon.com/directoryservice/latest/devguide/API_RadiusSettings.html

ValidateFunc: validation.IntBetween(0, 256),

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [inline aws/resource_aws_directory_service_directory.go:202]: `ValidateFunc: validation.IsPortNumber,`

File: aws/structure.go
Comment (DrFaust92) [inline aws/structure.go:1717]: use `flattenStringSet()`

File: aws/structure.go
Comment (DrFaust92) [inline aws/structure.go:1713]: use relevant aws.StringValue/Int64Value wrappers

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
Some comments to move this along a bit

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go]: done

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go]: done

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go:204]: done

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go:215]: done

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go:220]: done

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go:212]: Doesn't seem possible to add ValidateFunc to sets.  
I get the error:

```
1 error occurred:
        	* resource aws_directory_service_directory: servers: ValidateFunc and
        ValidateDiagFunc are not yet supported on lists or sets.
```

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go:199]: done

File: aws/structure.go
Comment (apagliara) [reply aws/structure.go]: done

File: aws/structure.go
Comment (apagliara) [reply aws/structure.go]: done

**Comment by apagliara** [comment]
@DrFaust92 Changes requested have been implemented.

Please could you review?  Many thanks

File: aws/resource_aws_directory_service_directory.go
Comment (DrFaust92) [reply aws/resource_aws_directory_service_directory.go:212]: it should be on the elem level:

```
Elem:     &schema.Schema{
Type: schema.TypeString,
ValidateFunc: validation.IntBetween(0, 256),
},
```

File: website/docs/r/directory_service_directory.html.markdown
Comment (DrFaust92) [inline website/docs/r/directory_service_directory.html.markdown:158]: lets add the vaild values for this (as this is a closed list) to make it easier on users

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
see comment on set validation above ^

lets also add a check on the basic test to check radius_setings is not being set
```
					resource.TestCheckResourceAttr(resourceName, "radius_settings.#", "0"),
```

Im not a maintainer so i cant approve this PR but ill try to validate tests later.

File: aws/resource_aws_directory_service_directory.go
Comment (apagliara) [reply aws/resource_aws_directory_service_directory.go]: Done, although it's a string, not an int.  Thanks

File: website/docs/r/directory_service_directory.html.markdown
Comment (apagliara) [reply website/docs/r/directory_service_directory.html.markdown]: done

**Comment by apagliara** [comment]
@DrFaust92 - changes have been made as requested and all tests pass.  Thanks

**Comment by tom-henderson** [comment]
@DrFaust92 just echoing @apagliara's comment above - can this PR please be assigned to a maintainer for approval?

**Comment by pointgenesis** [comment]
I am currently in need of this functionality. Please review for inclusion into the base product.

**Comment by rhysjtevans** [comment]
This would totally solve a horrid null_resource hack - any idea when it could be reviewed?

**Comment by stevesimpson418** [comment]
Echoing previous comments, this functionality would be of great use for anyone using AWS Directory Service (AWS Client VPN too) and needing MFA support 👍

**Comment by rtcsgit** [comment]
Also echoing previous comments.  Definitely would like this to be reviewed as we need this functionality!

**Comment by zhelding** [comment]
Pull request #21306 has significantly refactored the AWS Provider codebase. As a result, most PRs opened prior to the refactor now have merge conflicts that must be resolved before proceeding.

Specifically, PR #21306 relocated the code for all AWS resources and data sources from a single `aws` directory to a large number of separate directories in `internal/service`, each corresponding to a particular AWS service. This separation of code has also allowed for us to simplify the names of underlying functions -- while still avoiding namespace collisions.

We recognize that many pull requests have been open for some time without yet being addressed by our maintainers. Therefore, we want to make it clear that resolving these conflicts _in no way affects the prioritization_ of a particular pull request. Once a pull request has been prioritized for review, the necessary changes will be made by a maintainer -- either directly or in collaboration with the pull request author.

For a more complete description of this refactor, including examples of how old filepaths and function names correspond to their new counterparts: please refer to issue #20000.

For a quick guide on how to amend your pull request to resolve the merge conflicts resulting from this refactor and bring it in line with our new code patterns: please refer to our [Service Package Refactor Pull Request Guide](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/service-package-pullrequest-guide.md).

**Comment by ewbankkit** [comment]
In order to ensure that any existing directory RADIUS settings managed outside of Terraform aren't affected by this functionality, we have decided to implement this as a new, separate resource `aws_directory_service_radius_settings`.
This has the added advantage of allowing this new resource to manage RADIUS settings for replicated directories (`aws_directory_service_region` - #25755) as [MFA is a Regional feature of AWS Managed Microsoft AD](https://docs.aws.amazon.com/directoryservice/latest/admin-guide/ms_ad_mfa.html).

##### TODO

Add `radius_settings` attribute to `aws_directory_service_directory` data source.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

**Comment by ewbankkit** [comment]
@apagliara Thanks for the contribution :tada: :clap:.

**Comment by ewbankkit** [comment]
Due to the complexity involved in setting up a RADIUS server to test against, we have not been able to test the positive case (only ending up in a `Failed` state) and the resource is provided "as-is".
Please report any issues that are found in live scenarios.

**Pull request merged**
